### ISSUE TITLE
INTERNAL error with ghdl --synth

### ISSUE THREAD

**Issue by mikey** [ghdl/ghdl#2048]
**Description**
I get an internal error with tip and this code

**Expected behaviour**
No INTERNAL ERROR

**How to reproduce?**
```
% cat test.vhdl
library ieee;
use ieee.all.common.workall;
entity icache is end;
% ghdl --synth --std=08 --no-formal  test.vhdl -e icache
sem_selected_by_all_name: cannot handle IIR_KIND_LIBRARY_DECLARATION (test.vhdl:1:9)

******************** GHDL Bug occurred ***************************
Please report this bug on https://github.com/ghdl/ghdl/issues
GHDL release: 3.0.0-dev (2.0.0.r143.g01c330faae) [Dunoon edition]
Compiled with GNAT Version: 11.2.1 20220127 (Red Hat 11.2.1-
Target: x86_64-redhat-linux
/home/mikey/src/microwatt2/
Command line:
ghdl --synth --std=08 --no-formal test.vhdl -e icache
Exception TYPES.INTERNAL_ERROR raised
Exception information:
raised TYPES.INTERNAL_ERROR : vhdl-errors.adb:30
Call stack traceback locations:
0x5cc343 0x7114b7 0x714863 0x70e8ee 0x71484b 0x7159b9 0x6ea6f2 0x6eaa5b 0x6ead6c 0x6eb2d4 0x6c1e52 0x6c2cbc 0x6c2f18 0x6c5b15 0x745dde 0x7467c7 0x656e45 0x7f31fb 0x40bff5 0x7f4654ab355e 0x7f4654ab360a 0x40a5c3 0xfffffffffffffffe
******************************************************************
```

```vhd :file: test.vhdl
library ieee;
use ieee.all.common.workall;
entity icache is end;
```

```sh :image: ghdl/ghdl:bullseye-mcode
ghdl --synth --std=08 --no-formal  test.vhdl -e icache
```

**Context**
Please, provide the following information:

- OS: Linux
- Origin:

  - [x] Built from sources: `commit 01c330faae`

If a `GHDL Bug occurred` block is shown in the log, please paste it here:

```
******************** GHDL Bug occurred ***************************
Please report this bug on https://github.com/ghdl/ghdl/issues
GHDL release: 3.0.0-dev (2.0.0.r143.g01c330faae) [Dunoon edition]
Compiled with GNAT Version: 11.2.1 20220127 (Red Hat 11.2.1-
Target: x86_64-redhat-linux
/home/mikey/src/microwatt2/
Command line:
ghdl --synth --std=08 --no-formal test.vhdl -e icache
Exception TYPES.INTERNAL_ERROR raised
Exception information:
raised TYPES.INTERNAL_ERROR : vhdl-errors.adb:30
Call stack traceback locations:
0x5cc343 0x7114b7 0x714863 0x70e8ee 0x71484b 0x7159b9 0x6ea6f2 0x6eaa5b 0x6ead6c 0x6eb2d4 0x6c1e52 0x6c2cbc 0x6c2f18 0x6c5b15 0x745dde 0x7467c7 0x656e45 0x7f31fb 0x40bff5 0x7f4654ab355e 0x7f4654ab360a 0x40a5c3 0xfffffffffffffffe
******************************************************************
```

**Additional context**
Found using @antonblanchard [microwatt](https://github.com/antonblanchard/microwatt/) and cvise to reduce.

**Comment by umarcor** [comment]
I can confirm the error with either mcode or LLVM using either VHDL 1993 or VHDL 2008.

**Comment by mikey** [comment]
Thanks!

**Comment by mikey** [comment]
This fixes the original bug but I can now hit the same problem with this code 
```
library ieee;
use ieee.std_logic_1164.all.common.workall;
entity icache is end;
```
I'm generating this error by the attached test.vhdl code and running [cvise](https://github.com/marxin/cvise). Here's how I'm doing it. This might be more useful than submitting each bug I hit individually. 
```
% cat ./ghdl-cvise.sh  
#!/bin/bash

ghdl --synth --std=08 --no-formal test.vhdl -e icache 2>&1 |grep "Exception TYPES.INTERNAL_ERROR"

exit $?
% cvise --timeout 5 ./ghdl-cvise.sh  test.vhdl
00:00:00 INFO ===< 3038894 >===
00:00:00 INFO running 4 interestingness tests in parallel
00:00:00 INFO INITIAL PASSES
00:00:00 INFO ===< IncludesPass >===

.............. SNIPPED ................

Runtime: 202 seconds
Reduced test-cases:

--- /home/mikey/src/microwatt/test.vhdl ---
library ieee;
use ieee.std_logic_1164.all.common.workall;
entity icache is end;

%
```
[test.vhdl.gz](https://github.com/ghdl/ghdl/files/8606305/test.vhdl.gz)

**Comment by tgingold** [comment]
Interesting.  I will try cvise.

A long time ago, I tried to find a tool to automatically generate files according to a grammar, but with errors.  I was not very successful.

**Comment by mikey** [comment]
Thanks @tgingold 

Here's another one I'm having issues with. Slightly different grep in the interestingness test for cvise.

Original test case is here [test.vhdl.gz](https://github.com/ghdl/ghdl/files/8608773/test.vhdl.gz)


```
% cat ghdl-cvise.sh
#!/bin/bash

ghdl --synth --std=08 --no-formal  test.vhdl -e icache 2>&1 |grep "netlists-memories.adb:1419"
exit $?

% cvise --timeout 10 --not-c ghdl-cvise.sh test.vhdl
00:00:00 INFO ===< 1834962 >===
00:00:00 INFO running 4 interestingness tests in parallel
00:00:00 INFO INITIAL PASSES
00:00:00 INFO ===< BlankPass >===
00:00:00 INFO (0.5%, 10115 bytes, 258 lines)

.............. SNIPPED ................

Runtime: 224 seconds
Reduced test-cases:

--- /home/mikey/src/microwatt/test.vhdl ---
                 package wishbone_types is     constant wishbone_data_bits : integer := 64;
                        end ;
                        library ieee;
                        use ieee.std_logic_1164.all;
                        package common is     constant REAL_ADDR_BITS : positive := 6;
                            subtype real_addr_t is std_ulogic_vector(REAL_ADDR_BITS - 1 downto 0);
                            function addr_to_real(addr: std_ulogic_vector(63 downto 0)) return real_addr_t;
                            type Fetch1ToIcacheType is record 	req: std_ulogic;
                                virt_mode : std_ulogic;
                        	nia: std_ulogic_vector(63 downto 0);
                                big_endian: std_ulogic;
                            end record;
                            type MmuToIcacheType is record         tlbld : std_ulogic;
                                tlbie : std_ulogic;
                                pte   : std_ulogic_vector(63 downto 0);
                            end record;
                        end ;
                        package body common is     function addr_to_real(addr: std_ulogic_vector(63 downto 0)) return real_addr_t is     begin         return addr(real_addr_t'range);
    end;
end common;
library ieee;
use ieee.std_logic_1164.all;
use work.common.all;
use work.wishbone_types.all;
entity icache is
    port (
        clk          : in std_ulogic;
        rst          : in std_ulogic;
        i_in         : in Fetch1ToIcacheType;
        m_in         : in MmuToIcacheType;
	stall_out    : out std_ulogic
        );
end entity icache;
architecture rtl of icache is
    function log2(i : natural) return integer is
        variable ret : integer := 0;
    begin
        return ret;
    end function;
    constant LINE_SIZE : positive := 64;
    constant ROW_SIZE  : positive := wishbone_data_bits / 8;
    constant NUM_LINES : positive := 2;
    constant NUM_WAYS  : positive := 4;
    constant TLB_SIZE : positive := 4;
    constant TLB_LG_PGSZ : positive := 2;
    constant ROW_PER_LINE  : natural := LINE_SIZE / ROW_SIZE;
    constant BRAM_ROWS     : natural := NUM_LINES * ROW_PER_LINE;
    constant LINE_OFF_BITS : natural := log2(LINE_SIZE);
    constant INDEX_BITS    : natural := log2(NUM_LINES);
    constant SET_SIZE_BITS : natural := LINE_OFF_BITS + INDEX_BITS;
    constant TAG_BITS      : natural := REAL_ADDR_BITS - SET_SIZE_BITS + 1;
    subtype row_t is integer range 0 to BRAM_ROWS-1;
    subtype index_t is integer range 0 to NUM_LINES-1;
    subtype way_t is integer range 0 to NUM_WAYS-1;
    subtype cache_tag_t is std_logic_vector(TAG_BITS-1 downto 0);
    constant TAG_RAM_WIDTH : natural := TAG_BITS * NUM_WAYS;
    subtype cache_tags_set_t is std_logic_vector(TAG_RAM_WIDTH-1 downto 0);
    type cache_tags_array_t is array(index_t) of cache_tags_set_t;
    subtype cache_way_valids_t is std_ulogic_vector(NUM_WAYS-1 downto 0);
    type cache_valids_t is array(index_t) of cache_way_valids_t;
    type row_per_line_valid_t is array(0 to ROW_PER_LINE - 1) of std_ulogic;
    signal cache_tags   : cache_tags_array_t;
    signal cache_valids : cache_valids_t;
    constant TLB_PTE_BITS : natural := 64;
    subtype tlb_index_t is integer range 0 to TLB_SIZE - 1;
    subtype tlb_pte_t is std_ulogic_vector(TLB_PTE_BITS - 1 downto 0);
    type tlb_ptes_t is array(tlb_index_t) of tlb_pte_t;
    signal itlb_ptes : tlb_ptes_t;
    type state_t is (IDLE, STOP_RELOAD, CLR_TAG, WAIT_ACK);
    type reg_internal_t is record
        state            : state_t;
        rows_valid       : row_per_line_valid_t;
    end record;
    signal r : reg_internal_t;
    signal req_index   : index_t;
    signal req_row     : row_t;
    signal req_tag     : cache_tag_t;
    signal tlb_req_index : tlb_index_t;
    signal real_addr     : real_addr_t;
    function get_tag(addr: real_addr_t; endian: std_ulogic) return cache_tag_t is
    begin
        return endian & addr;
           end;
           function read_tag(way: way_t;
       tagset: cache_tags_set_t) return cache_tag_t is     begin 	return tagset((way+1) * TAG_BITS - 1 downto way * TAG_BITS);
           end;
       begin     itlb_lookup : process(all)         variable pte : tlb_pte_t;
           begin         pte := itlb_ptes(tlb_req_index);
               if i_in.virt_mode = '1' then             real_addr <= pte(REAL_ADDR_BITS - 1 downto TLB_LG_PGSZ) &                          i_in.nia(TLB_LG_PGSZ - 1 downto 0);
               else             real_addr <= addr_to_real(i_in.nia);
               end if;
           end process;
           itlb_update: process(clk, rst)         variable wr_index : tlb_index_t;
           begin 	if rst = '1' then 	    itlb_ptes <= (others => (others => '0'));
               elsif rising_edge(clk) then             if m_in.tlbie = '1' then             elsif m_in.tlbld = '1' then                 itlb_ptes(wr_index) <= m_in.pte;
                   end if;
               end if;
           end process;
           icache_comb : process(all) 	variable is_hit  : std_ulogic;
           begin         req_tag <= get_tag(real_addr, i_in.big_endian);
       	for i in way_t loop 	    if i_in.req = '1' and                 (cache_valids(req_index)(i) = '1' or                  (r.state = WAIT_ACK and                   r.rows_valid(req_row mod ROW_PER_LINE) = '1')) then 		if read_tag(i, cache_tags(req_index)) = req_tag then 		    is_hit := '1';
       		end if;
       	    end if;
       	end loop;
       	stall_out <= is_hit;
           end process;
       end;

% ghdl --synth --std=08 --no-formal  test.vhdl -e icache 2>&1 |tail -15

******************** GHDL Bug occurred ***************************
Please report this bug on https://github.com/ghdl/ghdl/issues
GHDL release: 3.0.0-dev (2.0.0.r151.g176b14533b) [Dunoon edition]
Compiled with GNAT Version: 11.3.1 20220421 (Red Hat 11.3.1-
Target: x86_64-redhat-linux
/home/mikey/src/microwatt2/
Command line:
ghdl --synth --std=08 --no-formal test.vhdl -e icache
Exception TYPES.INTERNAL_ERROR raised
Exception information:
raised TYPES.INTERNAL_ERROR : netlists-memories.adb:1419
Call stack traceback locations:
0x54b62a 0x54b7dc 0x54b935 0x54b903 0x54f755 0x681bb6 0x6b024b 0x6b0296 0x681b2c 0x7469a2 0x656e45 0x7f3323 0x40bff5 0x7f0d964e1f1e 0x7f0d964e1fce 0x40a5c3 0xfffffffffffffffe
******************************************************************
%
```

**Comment by tgingold** [comment]
This one is a different issue, as the crash appears in the memory inference.

**Comment by mikey** [comment]
Do you want me to raise a new issue on github?

**Comment by mikey** [comment]
@tgingold did you manage to get anywhere with this last issue?

**Comment by tgingold** [comment]
Yes, it's a crash in the memory inference code.  There are already a few issues about that, so I have to dig.

**Comment by tgingold** [comment]
Yes please.  But I don't plan to work on it immediately.
